src/parser/jobs/rpr/modules/HarvestMoon.tsx

```diff
@@ -35,37 +36,60 @@ export class HarvestMoon extends Analyser {
-		const ADJUSTED_CAST_TIME = this.data.actions.SOULSOW.castTime + SOULSOW_BUFFER
+	private getExpectedUses(): number {
+		return this.invulWindows()
+			.filter(this.canChargeMoon.bind(this))
+			.length + 1
+	}
 
-		// get the earliest unable to act window that falls within the provided inputWindow
-		const unableToActWindow = this.unableToAct.getWindows(inputWindow)[0]
+	private invulWindows(): Window[] {
+		return this.invulnerability.getWindows().filter(this.longerThanChargeTime.bind(this))
+	}
 
-		// the window will be undefined if there are no unable to act windows left before the end of the inputWindow
-		if (unableToActWindow == null) {
-			// True if there are ADJUSTED_CAST_TIME milliseconds or more remaining in the inputWindow
-			return inputWindow.end - inputWindow.start >= ADJUSTED_CAST_TIME
-		}
+	private canChargeMoon(invulWindow: Window): boolean {
+		return this.unableToActWindows(invulWindow)
+			.reduce(this.toActableWindows, [] as Window[])
+			.some(this.longerThanChargeTime.bind(this))
+	}
 
-		// return true if there are ADJUSTED_CAST_TIME milliseconds between the beginning of the window being checked and the beginning of an unableToAct window
-		if (unableToActWindow.start - inputWindow.start >= ADJUSTED_CAST_TIME) {
-			return true
+	// Reduce function for calculating interstitial windows that assumes that there are at least 2 windows.
+	private toActableWindows(acc: Window[], window: Window, idx: number, arr: Window[]): Window[] {
+		if (idx === arr.length - 1) {
+			return acc
 		}
-
-		// recurse the method, shrinking the window to the space between the end of the unable to act window and the end of the input window.
-		return this.canChargeMoon({start: unableToActWindow.end, end: inputWindow.end})
+		return acc.concat({
+			start: window.end,
+			end: arr[idx + 1].start,
+		})
 	}
 
-	private getExpectedUses(): number {
-		const ADJUSTED_CAST = this.data.actions.SOULSOW.castTime + SOULSOW_BUFFER
-		const invulnWindows = this.invulnerability.getWindows().filter((window) => window.end - window.start >=  ADJUSTED_CAST)
+	// Gets unableToAct windows and adds dummy windows at start and end if needed
+	// to calculate the actable Windows. Guarantees at least 2 windows.
+	private unableToActWindows(invulWindow: Window): Window[] {
+		const unactableWindows = this.unableToAct.getWindows(invulWindow) as Window[]
 
-		// will only run the window filter if there is a non-zero amount of unableToAct time during the fight.
-		if (this.unableToAct.getDuration({start: this.parser.pull.timestamp, end: this.parser.pull.timestamp + this.parser.pull.duration}) > 0) {
-			return invulnWindows.filter(window => this.canChargeMoon(window)).length + 1
+		// Edge case where the unableToAct window is larger than the invulWindow.
+		if (this.unableToActWindows.length === 1
+			&& unactableWindows[0].start < invulWindow.start
+			&& unactableWindows[0].end > invulWindow.end) {
```

Comment:
This is quite an edge case, is there any way to simplify this?

Reply:
Fixed this a little bit but it's kind of the same. When looking at time intervals, they can overlap in 4 ways:
```
1: Overlap at the start of "B" and end of "A"
AAAAA
    BBBBBBBBBBBBB
 2. Overlap at the end of "B" and start of "A"
        AAAAA
 BBBBB
 3. "A" is a subset of "B"
     AAA
BBBBBBBBB
 4. "B" is a subset of "A"
   AAAAAAAAA
      BBB
```

The conditionals here handle all 4 of these situations, and the edge case is 4, where "A" is the unactable window and "B" is the invulWindow. I think this probably won't ever happen in a real fight but it should be handled anyway.